### PR TITLE
Nucleo f072

### PR DESCRIPTION
**Pull request by jia200x**
Hi!

Added support for Nucleo F072 board. I'm having problems with OpenOCD for Mac OS X (I can flash only once, then stops working). Seems to work in Linux.

### CODE REVIEW

**Comment by haukepetersen** [comment]
looks good and works as expected. Could you please rebase and squash your commits? thanks

**Comment by jia200x** [comment]
Rebased and squashed.
Cheers!

**Comment by haukepetersen** [comment]
hm, seems like something went wrong. I am still seeing a number of commits that don't belong here..

**Comment by jia200x** [comment]
Yes, very weird... I have tried several times to rebase, but is not working
as I'm expecting. I think it's faster to create a new branch, add files and
change the variable in vendor header file and PR again. I think I'll do
that, is ok?

2016-03-02 16:08 GMT+01:00 Hauke Petersen <notifications@github.com>:

**Comment by jia200x** [comment]
Ok, fixed.
Could you test it again just for checking everything went right?

Thank you!

**Comment by jia200x** [comment]
It's seems now is OK

**Comment by haukepetersen** [comment]
looks good and works -> ACK once Travis is green.

**Comment by jia200x** [comment]
Cool!
Thank you.

2016-03-03 14:35 GMT+01:00 Hauke Petersen <notifications@github.com>:

**Comment by OlegHahm** [comment]
There are some Travis complaints:
```
boards/nucleo-f072/board.c:45: new blank line at EOF.

boards/nucleo-f072/include/board.h:68: new blank line at EOF.

boards/nucleo-f072/include/periph_conf.h:109: new blank line at EOF.

cpu/stm32f0/include/cpu_conf.h:33: trailing whitespace.

+ 

cpu/stm32f0/include/stm32f072xb.h:1: trailing whitespace.

+/**^M

cpu/stm32f0/include/stm32f072xb.h:2: trailing whitespace.

+  *****************************************************************************

cpu/stm32f0/include/stm32f072xb.h:3: trailing whitespace.

+  * @file    stm32f072xb.h^M

cpu/stm32f0/include/stm32f072xb.h:4: trailing whitespace.

+  * @author  MCD Application Team^M

cpu/stm32f0/include/stm32f072xb.h:5: trailing whitespace.

+  * @version V2.2.3^M

cpu/stm32f0/include/stm32f072xb.h:6: trailing whitespace.

+  * @date    29-January-2016^M

cpu/stm32f0/include/stm32f072xb.h:7: trailing whitespace.

+  * @brief   CMSIS Cortex-M0 Device Peripheral Access Layer Header File. ^M

cpu/stm32f0/include/stm32f072xb.h:8: trailing whitespace.

+  *          This file contains all the peripheral register's definitions, bits

cpu/stm32f0/include/stm32f072xb.h:9: trailing whitespace.

+  *          definitions and memory mapping for STM32F0xx devices.            

cpu/stm32f0/include/stm32f072xb.h:10: trailing whitespace.

+  *            ^M

cpu/stm32f0/include/stm32f072xb.h:11: trailing whitespace.

+  *          This file contains:^M

cpu/stm32f0/include/stm32f072xb.h:12: trailing whitespace.

+  *           - Data structures and the address mapping for all peripherals^M

cpu/stm32f0/include/stm32f072xb.h:13: trailing whitespace.

+  *           - Peripheral's registers declarations and bits definition^M

cpu/stm32f0/include/stm32f072xb.h:14: trailing whitespace.

+  *           - Macros to access peripheral<92>s registers hardware^M

cpu/stm32f0/include/stm32f072xb.h:15: trailing whitespace.

+  *  ^M

cpu/stm32f0/include/stm32f072xb.h:16: trailing whitespace.

+  *****************************************************************************

cpu/stm32f0/include/stm32f072xb.h:17: trailing whitespace.

+  * @attention^M

:

```

**Comment by OlegHahm** [comment]
It also needs some blacklisting for several applications.

**Comment by haukepetersen** [comment]
The Travis complains come from dos line endings in the vendor header file, so just convert them to unix ones...

**Comment by jia200x** [comment]
Hi!

Oops sorry, I forgot to finish this.
What kind of blacklisting do this model need?

I'm fixing it

2016-03-16 10:17 GMT+01:00 Hauke Petersen <notifications@github.com>:

**Comment by OlegHahm** [comment]
E.g.
```
gnrc_border_router:nucleo-f072:
Building application "gnrc_border_router" for "nucleo-f072" with MCU "stm32f0".
/usr/bin/../lib/gcc/arm-none-eabi/4.9.3/../../../../arm-none-eabi/bin/ld: /home/travis/build/RIOT-OS/RIOT/examples/gnrc_border_router/bin/nucleo-f072/gnrc_border_router.elf section `.bss' will not fit in region `ram'
/usr/bin/../lib/gcc/arm-none-eabi/4.9.3/../../../../arm-none-eabi/bin/ld: region `ram' overflowed by 3424 bytes
collect2: error: ld returned 1 exit status
make: *** [all] Error 1
```
this means, that you will have to add the board to the `BOARD_INSUFFICIENT_MEMORY` list in `examples/gnrc_border_router/Makefile`.

**Comment by jia200x** [comment]
Got it.
Thank you!

2016-03-16 11:54 GMT+01:00 Oleg Hahm <notifications@github.com>:

**Comment by jia200x** [comment]
Done.

I blacklisted the examples, but make command compiles everything and throws
"region 'ram' overflow". Is this a normal behavior?

Cheers!

2016-03-16 11:55 GMT+01:00 Jose Alamos <jialamos@uc.cl>:

**Comment by jia200x** [comment]
Hi.

Failed again. The examples are not being build. I blacklisted some of them,
but make still tries to build them. It's normal?

Also don't know what happened with the vendor header file, It seems I
commited the old version. I'll fix it and try again.

2016-03-16 13:07 GMT+01:00 Jose Alamos <jialamos@uc.cl>:

**Comment by OlegHahm** [comment]
Somehow you seemed to have reverted the blacklisting in https://github.com/jia200x/RIOT/commit/06dbcdf691ffe4b7fa0d2f8a208f74babd08ec33

**Comment by jia200x** [comment]
Yes, I noticed it.
Btw, for a strange reason the upstream master is not working for me. I
tried to merge because the code was not running in CI, but now it's not
working. I even tried a new clone of the RIOT repository, without luck.

Could be an issue with Xcode, I will try to fix it tomorrow.

**Comment by haukepetersen** [comment]
Looks good, but the paradigm for handling the on-board LED has changed recently. So you simply need to include `#include "board_common.h"` in the board.h file and change the LED initialization code in the board.c to 
```c
    /* initialize the boards LED */
    gpio_init(LED0_PIN, GPIO_OUT);
```
Also you need to include the `nucleo-common` module in the board Makefiles, just look at one of the existing nucleo boards...

**Comment by haukepetersen** [comment]
not critical for the release, so set the milestone to 2016.07

**Comment by jia200x** [comment]
Ok.

Anyway, I will finish it this week.
Cheers!

2016-03-29 17:35 GMT+02:00 Hauke Petersen <notifications@github.com>:

**Comment by haukepetersen** [comment]
As soon as you get to it. If you get it done by tomorrow, it will make the release (which would be great!).

**Comment by jia200x** [comment]
Done. Sorry for the delay, I had to edit vendor file and add DAC info.
Hope it works and could be introduced in the release.
Cheers!

**Comment by jia200x** [comment]
I had an error due white spaces. Just fixed.

Code is running :)

**Comment by OlegHahm** [comment]
I'm somewhat scared by the number of added lines, but since most of them are in the (bloated) STM CPU header, I guess this is fine. Additionally this PR introduces a new feature - so it cannot break much.

File: boards/nucleo-f072/include/board.h [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+ * @name Macros for controlling the on-board LEDs.
+ * @{
+ */
+#define LED_RED_ON
+#define LED_RED_OFF
+#define LED_RED_TOGGLE
+
+#define LED_GREEN_ON        (LED_GREEN_PORT->BSRR = (1 << LED_GREEN_PIN))
+#define LED_GREEN_OFF       (LED_GREEN_PORT->BSRR = ((1 << LED_GREEN_PIN) << 16))
+#define LED_GREEN_TOGGLE    (LED_GREEN_PORT->ODR ^= (1 << LED_GREEN_PIN))
```
Comment (OlegHahm) [inline boards/nucleo-f072/include/board.h]: needs to be adapted to #5045.

Comment (jia200x) [inline boards/nucleo-f072/include/board.h]: I forgot to remove these lines.
Just done.

**Comment by jia200x** [comment]
Beside of the number of lines, the header files changed common variable names and added some legacy variables that breaks compiling. I had to remove the #define DAC1 line due to the new implementation of DAC.

I will be aware if breaks again with a new change.

File: boards/nucleo-f072/include/periph_conf.h [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+#define UART_1_ISR          isr_usart1
+/* UART 1 pin configuration */
+#define UART_1_PORT         GPIOB
+#define UART_1_PORT_CLKEN() (RCC->AHBENR |= RCC_AHBENR_GPIOBEN)
+#define UART_1_RX_PIN       7
+#define UART_1_TX_PIN       6
+#define UART_1_AF           0
+/** @} */
+
+#define ADC_CONFIG {            \
```
Comment (haukepetersen) [inline boards/nucleo-f072/include/periph_conf.h]: missing doxygen for the ADC block

File: cpu/stm32f0/ldscripts/stm32f072rb.ld [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @addtogroup      cpu_stm32f0
+ * @{
+ *
+ * @file
+ * @brief           Memory definitions for the STM32F091RC
```
Comment (haukepetersen) [inline cpu/stm32f0/ldscripts/stm32f072rb.ld]: this is the linkderscript for the `STM32F072rb`, right?

**Comment by haukepetersen** [comment]
I don't quite understand, why you removed the DAC1 from the vendor header. The device clearly has a DAC included, so you should not remove it. This error should be solved in a different way.

**Comment by jia200x** [comment]
Hi.

According to the vendor header file, DAC was defined in the same way as
DAC1 due to legacy reasons. When the DAC1 was present, dac.c file from
stm32_common [1] failed at line 61. (RCC_APB1ENR_DACEN is defined, not
RCC_APB1ENR_DAC1EN).


That was the reason why I removed DAC1.
Should I instead define RCC_APB1ENR_DAC1EN?



2016-03-31 17:26 GMT+02:00 Hauke Petersen <notifications@github.com>:

**Comment by jia200x** [comment]
[1]
https://github.com/RIOT-OS/RIOT/blob/d411387bec067098f2692de8332d07af82973137/cpu/stm32_common/periph/dac.c

2016-03-31 17:32 GMT+02:00 Jose Alamos <jialamos@uc.cl>:

**Comment by haukepetersen** [comment]
Oh, I see, then never mind the comment about the `DAC`...

**Comment by haukepetersen** [comment]
would just be nice if you could squash your commits a little bit, I would suggest 1 commit for adding the CPU, 1 for editing the vendor header, 1 for adding the board, and 1 for the blacklisting...

**Comment by jia200x** [comment]
I'm on it. Had some conflicts with rebase, but I'm fixing them.

**Comment by haukepetersen** [comment]
looks good, just one more thing that Murdock complains about, there are some whitespace errors in the board files:
```

BUILD OUTPUT of group static-tests:
Current branch HEAD is up to date.
Current branch HEAD is up to date.
boards/nucleo-f072/board.c:34: new blank line at EOF.
boards/nucleo-f072/include/board.h:48: new blank line at EOF.
boards/nucleo-f072/include/periph_conf.h:136: new blank line at EOF.
boards/nucleo-f072/board.c:34: new blank line at EOF.
boards/nucleo-f072/include/board.h:48: new blank line at EOF.
boards/nucleo-f072/include/periph_conf.h:136: new blank line at EOF.
ERROR: This change introduces new whitespace errors
```

**Comment by jia200x** [comment]
Done. It seems it was introducing when I tried to solve conflicts of rebase.

**Comment by haukepetersen** [comment]
Alright, Murdock is green, code looks good, I'll make an exception here and merge this (as feature freeze is very young and the release branch is not created yet...) -> ACK and go

**Comment by jia200x** [comment]
:)

**Comment by haukepetersen** [comment]
nice work by the way!

**Comment by jia200x** [comment]
Thank you!
Helped me a lot to understand PR<->CI-> ACK loop.

I will check for missing documentation about this.
Cheers!

**Pull request merged**